internal/sqldb/sqlquery/observation.go

```diff
@@ -0,0 +1,170 @@
+	for rows.Next() {
+		var entity, variable, date, provenance, unit, scalingFactor, measurementMethod, observationPeriod, properties string
+		var value float64
+		if err := rows.Scan(&entity, &variable, &date, &value, &provenance, &unit, &scalingFactor, &measurementMethod, &observationPeriod, &properties); err != nil {
+			return nil, err
+		}
+		observation := &pb.PointStat{
+			Date:  date,
+			Value: proto.Float64(value),
+		}
+
+		facetId, facet := toFacet(cachedProvenances, provenance, unit, scalingFactor, measurementMethod, observationPeriod, properties)
+		tsKey := byFacetKey{variable: variable, entity: entity, facetId: facetId}
+		tsValue := intermdiate.byFacet[tsKey]
+		if tsValue == nil {
+			tsValue = &byFacetValue{facetId: facetId, facet: facet}
+			intermdiate.byFacet[tsKey] = tsValue
+		}
+		tsValue.observations = append(tsValue.observations, observation)
+	}
+
+	return &intermdiate, rows.Err()
+}
+
+func newObservationResponse(variables []string) *pbv2.ObservationResponse {
+	result := &pbv2.ObservationResponse{
+		ByVariable: map[string]*pbv2.VariableObservation{},
+		Facets:     map[string]*pb.Facet{},
+	}
+	for _, variable := range variables {
+		result.ByVariable[variable] = &pbv2.VariableObservation{
+			ByEntity: map[string]*pbv2.EntityObservation{},
+		}
+	}
+	return result
+}
+
+// toFacet returns a facet ID and facet based on the specific observation properties.
+// Note that the "properties" argument which includes custom properties is not currently used.
+// But it can be used in the future, if we add a provision of custom properties to the Facet proto.
+func toFacet(cachedProvenances map[string]*pb.Facet, provenance, unit, scalingFactor, measurementMethod, observationPeriod, properties string) (string, *pb.Facet) {
+	cachedFacet := cachedProvenances[provenance]
+	if cachedFacet == nil {
+		cachedFacet = &pb.Facet{}
+	}
+	facet := proto.Clone(cachedFacet).(*pb.Facet)
+	facet.Unit = unit
+	facet.ScalingFactor = scalingFactor
+	facet.MeasurementMethod = measurementMethod
+	facet.ObservationPeriod = observationPeriod
+	return util.GetFacetID(facet), facet
+}
+
+func getObservationsSQLQuery(variables []string,
+	entities []string, queryDate string) string {
+	entitiesStr := "'" + strings.Join(entities, "', '") + "'"
+	variablesStr := "'" + strings.Join(variables, "', '") + "'"
+	query := fmt.Sprintf(
+		`
+			SELECT entity, variable, date, value, provenance, unit, scaling_factor, measurement_method, observation_period, properties FROM observations
+			WHERE entity IN (%s)
+			AND variable IN (%s)
+			AND value != ''
+		`,
+		entitiesStr,
+		variablesStr,
+	)
+	if queryDate != "" && queryDate != shared.LATEST {
+		query += fmt.Sprintf("AND date = (%s) ", queryDate)
+	}
+	query += "ORDER BY date ASC;"
+	return query
+}
+
+// The internal structs below are for generating an intermediate response from the SQL response to simplify generating the final ObservationResponse.
+type intermediateObservationResponse struct {
```

Comment:
This name is too long for a type, also since it just have one field, why not have a type alias for map[byFacetKey]*byFacetValue and call it `facetMap`?

Reply:
Actually, I needed to add another field to maintain a consistent order.